### PR TITLE
feat(doc): bao test framework

### PR DESCRIPTION
**Pull request by Diogo21Costa**
### Description
This PR introduces a description of a testing framework for the Bao hypervisor project. The framework enables developers to conduct unit tests on system components and assess interactions between different layers by leveraging a Python-based test framework API that handles build system interaction, test setup generation, test runner invocation, and result logging.

### Changes Made

- Provided an overview of the framework's concepts, configuration options, and Nix recipes.
- Explained the test definition process using the BAO_TEST macro and integrated assertion functions.
- Recommended directory structure and integration instructions for the build system.
- Included instructions for writing tests, running tests locally, and setting up CI with GitHub Actions.

The documentation provides a comprehensive guide on the purpose, features, and usage of the framework. It includes step-by-step instructions for integrating the framework and utilizing it effectively for testing.
Please review this PR and provide your feedback.

### CODE REVIEW

File: source/development/testing.rst [hunk 1]
```diff
@@ -1,2 +1,416 @@
...
+developers to conduct unit tests on various system components (such as the
+hypervisor, virtual machine (VM), and virtual machine monitor (VMM)), as well
+as assess interactions between different components/layers.
+
+Test Framework
+---------------
+
+Nomenclature
+*************
+- **MUT**: Module Under Test;
```
[the remark this answers is not in the document]
Comment (ESCristiano) [reply source/development/testing.rst]: I have removed the Nomenclature section and added the terms to the glossary as suggested.

File: source/development/testing.rst [hunk 2]
```diff
@@ -1,2 +1,416 @@
...
+
+Overview
+***********
+
+The test framework API is a Python tool that acts as the control center for the
+ testing process. It performs the following tasks:
```
[the remark this answers is not in the document]
Comment (ESCristiano) [reply source/development/testing.rst]: Whitespace removed!

File: source/development/testing.rst [hunk 3]
```diff
@@ -1,2 +1,416 @@
...
+status (pass/fail) and any error messages.
+
+.. figure:: img/framework-overview.png
+    :align: center
+    :name: bao-test-fig
+
+    Bao test framework overview.
+
+Overall, the test framework API serves as an intermediary between the build
+ system, the test setup generation process, and the test runner.
```
[the remark this answers is not in the document]
Comment (ESCristiano) [reply source/development/testing.rst]: Whitespace removed!

File: source/development/testing.rst [hunk 4]
```diff
@@ -1,2 +1,416 @@
...
+
+Test Framework
+---------------
+
+Nomenclature
+*************
+- **MUT**: Module Under Test;
+
+Overview
+***********
```
[the remark this answers is not in the document]
Comment (ESCristiano) [reply source/development/testing.rst]: Heading updated!

File: source/development/testing.rst [hunk 5]
```diff
@@ -1,2 +1,416 @@
...
+     platform: ["qemu-aarch64-virt"]
+   steps:
+     - uses: actions/checkout@v2
+      with:
+       submodules: recursive
+     - run: make test-framework name_x.dts
+
+
+Test Configuration File
+***********************
```
[the remark this answers is not in the document]
Comment (ESCristiano) [reply source/development/testing.rst]: Heading updated!

File: source/glossary.rst [hunk 6]
```diff
@@ -34,3 +34,6 @@ Glossary
         invited by a maintainer to oversee the development of one or multiple
         subsystems in a repository. More information
         :ref:`here <contrib_roles>`.
+
+    MUT
+        Module Under Test. More information :ref:`here <bao_tests>`.
```
Comment (danielRep) [inline source/glossary.rst:39]: Try to give a proper definition to the term. Something as "...refers to the specific component of the system that is currently being subjected to testing to ensure its proper functionality and correctness. " Or something more clear and specific of the testing framework.

**Comment by josecm** [comment]
> I believe you forked your feature branch from the wrong branch. The first commit, which is not part of this PR, is not in the main branch.

I've remove that commit and rebased this on main.

File: source/glossary.rst (hunk 6 above)
Comment (danielRep) [reply source/glossary.rst:61]: Besides, a newline must be in-place between terms or this will trigger an error on the build.

**Comment by josecm** [comment]
I believe the documentation introduced in this PR will be developed in [wip/testing](https://github.com/bao-project/bao-docs/tree/wip/testing), right @Diogo21Costa? If so, let's close this.

**Comment by Diogo21Costa** [comment]
> I believe the documentation introduced in this PR will be developed in [wip/testing](https://github.com/bao-project/bao-docs/tree/wip/testing), right @Diogo21Costa? If so, let's close this.

Yes, you're right @josecm.

**Pull request closed without merging**